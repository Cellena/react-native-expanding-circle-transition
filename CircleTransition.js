import React, { Component, PropTypes } from 'react'
import { Easing, Dimensions, Animated } from 'react-native'
const { width, height } = Dimensions.get('window')
const EXPAND = 1
const SHRINK = 0

class CircleTransition extends Component {
  constructor (props) {
    super(props)
    this.state = {
      visible: false,
      scale: new Animated.Value(SHRINK)
    }
  }

  start (callback) {
    const {expand} = this.props
    this.setModalVisible(true, expand, () => {
      this.animate(expand, callback)
    })
  }

  animate (expand, callback) {
    let toValue = expand ? EXPAND : SHRINK
    const { easing } = this.props
    Animated.timing(this.state.scale, {
      toValue: toValue,
      duration: this.props.duration,
      easing: easing
    }).start(() => {
      callback()
      this.setModalVisible(false, expand)
    })
  }

  setModalVisible (visible, expand, callback) {
    let fromValue = expand ? SHRINK : EXPAND
    this.setState({
      visible: visible,
      scale: new Animated.Value(fromValue)
    }, callback)
  }

  getLeftPosition (position) {
    const {size, customLeftMargin} = this.props
    const halfSize = size / 2
    const halfWidth = width / 2
    let marginHorizontalTopLeft = -halfSize

    switch (position) {
      case 'center':
      case 'top':
      case 'bottom':
        return marginHorizontalTopLeft + halfWidth

      case 'topRight':
      case 'bottomRight':
      case 'right':
        return marginHorizontalTopLeft + width

      case 'custom':
        return marginHorizontalTopLeft + customLeftMargin

      default:
        return marginHorizontalTopLeft
    }
  }

  getTopPosition (position) {
    const {size, customTopMargin} = this.props
    const halfSize = size / 2
    const halfHeight = height / 2
    let marginVerticalTopLeft = -halfSize

    switch (position) {
      case 'center':
      case 'left':
      case 'right':
        return marginVerticalTopLeft + halfHeight

      case 'bottomLeft':
      case 'bottomRight':
      case 'bottom':
        return marginVerticalTopLeft + height

      case 'custom':
        return marginVerticalTopLeft + customTopMargin

      default:
        return marginVerticalTopLeft
    }
  }

  render () {
    const {scale, visible} = this.state
    const {size, color, position, zIndex} = this.props
    let topPosition = this.getTopPosition(position)
    let leftPosition = this.getLeftPosition(position)
    if (visible) {
      return (
        <Animated.View style={{
          position: 'absolute',
          backgroundColor: color,
          zIndex: zIndex,
          top: topPosition,
          left: leftPosition,
          width: size,
          height: size,
          borderRadius: size / 2,
          transform: [{
            scale: scale
          }]
        }} />
      )
    } else {
      return null
    }
  }
}

CircleTransition.propTypes = {
  color: PropTypes.string,
  size: PropTypes.number,
  duration: PropTypes.number,
  position: PropTypes.oneOf([
    'topLeft',
    'topRight',
    'bottomLeft',
    'bottomRight',
    'center',
    'left',
    'right',
    'top',
    'bottom',
    'custom'
  ]),
  customLeftMargin: PropTypes.number,
  customTopMargin: PropTypes.number,
  expand: PropTypes.bool,
  easing: PropTypes.func,
  zIndex: PropTypes.number
}

CircleTransition.defaultProps = {
  color: 'orange',
  size: height * 3,
  duration: 800,
  position: 'topLeft',
  expand: true,
  customLeftMargin: 0,
  customTopMargin: 0,
  easing: Easing.linear,
  zIndex: 999
}
export default CircleTransition
