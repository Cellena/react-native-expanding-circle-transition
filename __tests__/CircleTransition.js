import 'react-native'
import React from 'react'
import CircleTransition from '../CircleTransition.js'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

const positionTest = (position) => {
  const tree = renderer.create(
    <CircleTransition
      color={'orange'}
      position={position}
      expand />
  ).toJSON()
  expect(tree).toMatchSnapshot()
}

it('renders correctly with default props', () => {
  const tree = renderer.create(
    <CircleTransition />
  )
})

it('renders correctly when position is center', () => {
  positionTest('center')
})

it('renders correctly when position is topRight', () => {
  positionTest('topRight')
})

it('renders correctly when position is topLeft', () => {
  positionTest('topLeft')
})

it('renders correctly when position is bottomLeft', () => {
  positionTest('bottomLeft')
})

it('renders correctly when position is bottomRight', () => {
  positionTest('bottomRight')
})
