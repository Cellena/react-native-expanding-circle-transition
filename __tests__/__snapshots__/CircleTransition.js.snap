exports[`test renders correctly when position is bottom 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  transparent={true}
  visible={false}>
  <View
    style={
      Object {
        "backgroundColor": "orange",
        "borderRadius": 2001,
        "height": 4002,
        "marginHorizontal": -1626,
        "marginVertical": -667,
        "transform": Array [
          Object {
            "scale": 0,
          },
        ],
        "width": 4002,
      }
    } />
</Modal>
`;

exports[`test renders correctly when position is bottomLeft 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  transparent={true}
  visible={false}>
  <View
    style={
      Object {
        "backgroundColor": "orange",
        "borderRadius": 2001,
        "height": 4002,
        "marginHorizontal": -2001,
        "marginVertical": -667,
        "transform": Array [
          Object {
            "scale": 0,
          },
        ],
        "width": 4002,
      }
    } />
</Modal>
`;

exports[`test renders correctly when position is bottomRight 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  transparent={true}
  visible={false}>
  <View
    style={
      Object {
        "backgroundColor": "orange",
        "borderRadius": 2001,
        "height": 4002,
        "marginHorizontal": -1251,
        "marginVertical": -667,
        "transform": Array [
          Object {
            "scale": 0,
          },
        ],
        "width": 4002,
      }
    } />
</Modal>
`;

exports[`test renders correctly when position is center 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  transparent={true}
  visible={false}>
  <View
    style={
      Object {
        "backgroundColor": "orange",
        "borderRadius": 2001,
        "height": 4002,
        "marginHorizontal": -1626,
        "marginVertical": -1334,
        "transform": Array [
          Object {
            "scale": 0,
          },
        ],
        "width": 4002,
      }
    } />
</Modal>
`;

exports[`test renders correctly when position is left 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  transparent={true}
  visible={false}>
  <View
    style={
      Object {
        "backgroundColor": "orange",
        "borderRadius": 2001,
        "height": 4002,
        "marginHorizontal": -2001,
        "marginVertical": -1334,
        "transform": Array [
          Object {
            "scale": 0,
          },
        ],
        "width": 4002,
      }
    } />
</Modal>
`;

exports[`test renders correctly when position is right 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  transparent={true}
  visible={false}>
  <View
    style={
      Object {
        "backgroundColor": "orange",
        "borderRadius": 2001,
        "height": 4002,
        "marginHorizontal": -1251,
        "marginVertical": -1334,
        "transform": Array [
          Object {
            "scale": 0,
          },
        ],
        "width": 4002,
      }
    } />
</Modal>
`;

exports[`test renders correctly when position is top 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  transparent={true}
  visible={false}>
  <View
    style={
      Object {
        "backgroundColor": "orange",
        "borderRadius": 2001,
        "height": 4002,
        "marginHorizontal": -1626,
        "marginVertical": -2001,
        "transform": Array [
          Object {
            "scale": 0,
          },
        ],
        "width": 4002,
      }
    } />
</Modal>
`;

exports[`test renders correctly when position is topLeft 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  transparent={true}
  visible={false}>
  <View
    style={
      Object {
        "backgroundColor": "orange",
        "borderRadius": 2001,
        "height": 4002,
        "marginHorizontal": -2001,
        "marginVertical": -2001,
        "transform": Array [
          Object {
            "scale": 0,
          },
        ],
        "width": 4002,
      }
    } />
</Modal>
`;

exports[`test renders correctly when position is topRight 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  transparent={true}
  visible={false}>
  <View
    style={
      Object {
        "backgroundColor": "orange",
        "borderRadius": 2001,
        "height": 4002,
        "marginHorizontal": -1251,
        "marginVertical": -2001,
        "transform": Array [
          Object {
            "scale": 0,
          },
        ],
        "width": 4002,
      }
    } />
</Modal>
`;
